src/safari-app-extension/AlpheiosReadingTools/AppDelegate.swift

```diff
@@ -117,9 +215,225 @@ class AppDelegate: NSObject, NSApplicationDelegate {
         // Insert code here to tear down your application
     }
     
+    func windowWillReturnUndoManager(window: NSWindow) -> UndoManager? {
+        // Returns the NSUndoManager for the application. In this case, the manager returned is that of the managed object context for the application.
+        return persistentContainer.viewContext.undoManager
+    }
+    
+    func applicationShouldTerminate(_ sender: NSApplication) -> NSApplication.TerminateReply {
+        // Save changes in the application's managed object context before the application terminates.
+        let context = persistentContainer.viewContext
+        
+        if !context.commitEditing() {
+            NSLog("\(NSStringFromClass(type(of: self))) unable to commit editing to terminate")
+            return .terminateCancel
+        }
+        
+        if !context.hasChanges {
+            return .terminateNow
+        }
+        
+        do {
+            try context.save()
+        } catch {
+            let nserror = error as NSError
+            
+            // Customize this code block to include application-specific recovery steps.
+            let result = sender.presentError(nserror)
+            if (result) {
+                return .terminateCancel
+            }
+            
+            let question = NSLocalizedString("Could not save changes while quitting. Quit anyway?", comment: "Quit without saves error question message")
```

Comment:
what changes are we trying to save here?

Reply:
This is to store all unsaved Core Data changes that are in memory but might not be saved to the hard drive. We do that before an application termination to make sure no in-memory changes will be lost.

Reply:
ok. Can you just add a comment to the file about that?

Reply:
Yes, will do that